### PR TITLE
Decorate class

### PR DESCRIPTION
**Pull request by felixDev22**
Hi There 👋,

**In this repo Learning objectives**

- [x] Implement composition (as an example of the design pattern).
- [x] Run a program using the command line.

**I have implemented the following**

- **Created a class Nameable.**
> Implement a method called correct_name that will raise a NotImplementedError.

- **Turned to Person class**
> Made sure that Person class inherits from Nameable
> Made sure that this class has a method correct_name implemented. And returns the name attribute.

- **Prepared base Decorator**
> Made sure that it inherits from Nameable.
> In the constructor assigned a nameable object from params to an instance variable.
> Implemented the correct_name method that returns the result of the correct_name method of the @nameable

- **Prepare CapitalizeDecorator and TrimmerDecorator**
> Created a class that inherits from the base Decorator class.
> Implemented a method correct_name that capitalizes the output of @nameable.correct_name.

- **For the TrimmerDecorator:**
Created a class that inherits from the base Decorator class.
Implemented a method correct_name that made sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it trim's the word.

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    raise NotImplementedError, 'Subclasses must implement the correct_name method'
```
Comment (christianbiring1) [inline nameable.rb]: [OPTIONAL]
- [x] To make this error rendering the class's name and method dynamically, you may consider saying 

```rb
raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
```

File: person.rb [hunk 2]
```diff
@@ -1,16 +1,20 @@
-class Person
+require_relative './nameable'
+
+class Person < Nameable
+  attr_reader :id
+  attr_accessor :name, :age
+
   def initialize(age, parent_permission: true, name: 'Unknown')
```
Comment (christianbiring1) [inline person.rb:7]: - [x] Kindly make sure, the user can have the possibility to give the `parent_permission` and `name` other values than the default ones and as you may notice the `app` is crushing when assign those variables as value

image


Hint:
```rb
def initialize(age, parent_permission = true, name = 'Unknown')
```

File: person.rb [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
-  # getter for id
-  attr_reader :id
-
-  # getter and setter for name and age
-  attr_accessor :name, :age
+  def correct_name
+    @name
+  end
 
   private
```
Comment (christianbiring1) [inline person.rb:15]: - [x] Kindly be aware that after the `private` keyword all the methods after it are private to that class. so kindly make sure after declaring the method you want to set as private, make sure you write `public` to make all the other following methods after that public again or you may consider writing all the private methods at the end of the class

File: trimmer_decorator.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require_relative './decorator'
+
+class TrimmerDecorator < Decorator
+  def correct_name
+    name = @nameable.correct_name
+    if name.length > 10
+      name.slice(0, 10)
+    else
+      name
```
Comment (christianbiring1) [inline trimmer_decorator.rb]: [OPTIONAL]
- [x] This code could be refactored and work the exact same way this way:
```rb
def
  @nameable.correct_name.length > 10 ? @nameable.correct_name[0..9] : @nameable.correct_name
end
```

**Review by christianbiring1 — CHANGES_REQUESTED** [review]
Hi @felixDev22 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
# REQUEST CHANGES :recycle: 


## Highlights
- [ ] All linters checks are green :heavy_check_mark: 
- [ ] Descriptive PR `title` and `summary` :+1: 

## Required Changes ♻️
- [ ] `KINDLY ALWAYS MAKE SURE YOU TEST YOUR CODE LOCALLY BEFRO YOU SUBMIT IT FOR A REVIEW` :muscle: 
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: trimmer_decorator.rb (hunk 4 above)
Comment (felixDev22) [reply trimmer_decorator.rb]: Thank you 😄  @christianbiring1 - The feedback is very informative, and I appreciate the clarity and wise advice. 👍 💯 
I have implemented your suggestion. Thanks again.

Comment (christianbiring1) [reply trimmer_decorator.rb]: Any time! @felixDev22 . Happy to hear that it was helpful

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @felixDev22👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**